R/mol_roi.R

```diff
@@ -0,0 +1,253 @@
+#' outside the roi extent. If false, occurrences are checked for whether they are
+#' located within the roi polygons or raster mask.
+#' @param maskValue numeric or character. Raster value corresponding to "outside"
+#' the region of interest. Ignored if roi is provided as a vector layer.
+#'
+#' @details This test identifies records outside a particular region of interest
+#' defined by a raster or vector layer
+#'
+#' @return A data.frame containing the column ".outside_roi". Compliant
+#' (TRUE) if coordinate is outside user defined roi; otherwise "FALSE".
+#'
+#' @importFrom sf st_bbox st_as_sf st_as_sfc st_within
+#' @importFrom stringr str_detect
+#' @importFrom terra crs rast geom vect
+#'
+#' @export mol_roi
+#'
+#' @examples
+#' x <- data.frame(
+#'   decimalLatitude = c(1, 2, 3, 5),
+#'   decimalLongitude = c(1, 2, 3, 4)
+#' )
+#'
+#' roi <- sf::st_as_sfc(sf::st_bbox(c(xmin = 0, xmax = 4, ymax = 4, ymin = 0), crs = 4326))
+#' 
+#' mol_roi(
+#'   data = x,
+#'   roi  = roi,
+#'   lat = "decimalLatitude",
+#'   lon = "decimalLongitude"
+#' )
+#'
+
+mol_roi <- 
+  function(data,
+           roi,
+           lat = "decimalLatitude",
+           lon = "decimalLongitude",
+           byExtOnly = FALSE,
+           maskValue = NA){
+    
+    ### Run checks
+    if (!is.data.frame(data)) {
+      stop(deparse(substitute(data)), " is not a data.frame", call. = FALSE)
+    }
+    
+    check_col(data, c(lat, lon))
+    
+    if(!any(c("character",
+              "sf",
+              "sfc",
+              "SpatialPolygons",
+              "RasterLayer",
+              "SpatRaster")) %in% class(roi)){
+      stop("The region of interest is not in a valid format.")
+    }
+    
+    ### converts coordinates columns to numeric
+    data <-
+      data %>%
+      dplyr::rowid_to_column("id_temp") %>%
+      dplyr::mutate(decimalLatitude = as.numeric(.data[[lat]]),
+                    decimalLongitude = as.numeric(.data[[lon]]),
+                    .outside_roi = TRUE) 
+    
+    ### Screen NAs
+    dataCoords %>% 
+      dplyr::select(dplyr::all_of(c("id_temp", lon, lat))) %>%
+      dplyr::filter(!is.na(.data[[lat]]), 
+                    !is.na(.data[[lon]]))
+    
+    if(nrow(dataCoords) == 0) stop("No valid coordinates.")
+    
+    ### Read file path
+    if(is.character(roi)){
+      
+      # check file exists
+      if(!file.exists(roi)) stop("The ROI filepath is invalid.")
+      
+      # check proper format
+      if(!any(str_detect(roi,
+                     c("\\.shp$", "\\.gpkg$", "\\.tif$")))){
+        stop("ROI input files must have '.shp', '.gpkg' or '.tif' file extensions.")
+      }
+      
+      # load roi
+      if(str_detect(roi, "\\.tif")){
+        roi <- terra::rast(roi)
+      } else{
+        roi <- st_read(roi)
+      }
+      
+    }
+    
+    ### Convert to sf/terra
+    if("SpatialPolygons" %in% .class2(roi)) roi <- st_as_sfc(roi)
+    if(class(roi) == "RasterLayer") roi <- rast(roi)
```

Comment:
There is an error here

if(class(roi) == "RasterLayer") roi <- rast(roi)
Error in if (class(roi) == "RasterLayer") roi <- rast(roi) : 
  the condition has length > 1

Maybe can be replaced by 
class(roi)[1] == "RasterLayer"